nvflare/private/fed/app/simulator/simulator_runner.py

```diff
@@ -0,0 +1,383 @@
+
+    def _deploy_apps(self, job_name, data_bytes, meta):
+        with tempfile.TemporaryDirectory() as temp_dir:
+            if os.path.exists(temp_dir):
+                shutil.rmtree(temp_dir)
+            os.mkdir(temp_dir)
+            unzip_all_from_bytes(data_bytes, temp_dir)
+            temp_job_folder = os.path.join(temp_dir, job_name)
+
+            app_server_root = os.path.join(self.simulator_root, "app_server")
+            for app_name, participants in meta.get(JobMetaKey.DEPLOY_MAP).items():
+                if len(participants) == 1 and participants[0].upper() == ALL_SITES:
+                    participants = ["server"]
+                    participants.extend([client for client in self.client_names])
+
+                for p in participants:
+                    if p == "server":
+                        app = os.path.join(temp_job_folder, app_name)
+                        shutil.copytree(app, app_server_root)
+                    else:
+                        app_client_root = os.path.join(self.simulator_root, "app_" + p)
+                        app = os.path.join(temp_job_folder, app_name)
+                        shutil.copytree(app, app_client_root)
+
+    def split_names(self, client_names: [], gpus: []):
+        split_names = []
+        for _ in gpus:
+            split_names.append([])
+        index = 0
+        for name in client_names:
+            split_names[index % len(gpus)].append(name)
+            index += 1
+        return split_names
+
+    def run(self):
+        try:
+            self.logger.info("Deploy and start the Server App.")
+            server_thread = threading.Thread(target=self.start_server_app, args=[])
+            server_thread.start()
+
+            # wait for the server app is started
+            while self.services.engine.engine_info.status != MachineStatus.STARTED:
+                time.sleep(1.0)
+
+            if self.args.gpu:
+                gpus = self.args.gpu.split(",")
+                split_client_names = self.split_names(self.client_names, gpus)
+                for index in range(len(gpus)):
+                    command = (
+                        sys.executable
+                        + " -m nvflare.private.fed.app.simulator.client_run_process -o "
+                        + self.args.workspace
+                        + " -c "
+                        + ",".join([name for name in split_client_names[index]])
+                        + " -g "
+                        + gpus[index]
+                        + " -p "
+                        + ",".join([str(i) for i in self.deployer.open_ports])
+                    )
+                    _ = subprocess.Popen(shlex.split(command, True), preexec_fn=os.setsid, env=os.environ.copy())
+            else:
+                client_runner = SimulatorClientRunner(self.args, self.client_names, self.deployer)
+                client_runner.run()
```

Comment:
In here you are using subprocess.Popen if self.args.gpu is active.

So inside your SimulatorClientRunner.run() you don't need to use subprocess.Popen anymore

Reply:
This is to split the clients onto different GPUs. Multi-GPU need to have sub-processes to run different clients.

Reply:
got it, so the self.args.gpu does NOT indicate whether to use GPU.

It actually means self.multi_gpu.
Then can we change the option to let user pass in multi_gpu?
And required users to also specified a GPU list if multi_gpu is enabled

Reply:
This also means either way we are going to use subprocess.

I suggest we always use subprocess Popen in here to run client_run_process.py (or we rename it as simulator_worker.py):

   1. If user not specify, we launch 1 client_run_process.py.
   2. If the user specify multi-GPU we launch several client_run_processes.py.

For the "SimulatorClientRunner.run()" function we can just run things directly.
We don't need to do another subprocess Popen there.

Reply:
This has been changed. Only ClientTaskWorker run with subprocess.